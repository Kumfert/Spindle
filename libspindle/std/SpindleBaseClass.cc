//
// SpindleBaseClass.cc  -- A cosmic base class for the Spindle framework
//
//  $Id: SpindleBaseClass.cc,v 1.2 2000/02/18 01:31:59 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_BASE_CLASS_H_
#include "spindle/SpindleBaseClass.h"
#endif

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static const ClassMetaData SpindleBaseClass_MetaData = {
  const_cast<char*>("SpindleBaseClass"), 
  sizeof( SpindleBaseClass ), 
  0, 
  SpindleBaseClass::createObject,
  0, 0
  };


static int UniqueIDNumber = 0;

const ClassMetaData* SpindleBaseClass::MetaData = &SpindleBaseClass_MetaData;

static const registry_init registry_init_dummy_class( &SpindleBaseClass_MetaData );

registry_init::registry_init( const ClassMetaData* metaData ) {
  SpindleSystem::registry().registerClass( metaData );
}

#ifdef __FUNC__
#undef __FUNC__
#endif

const ClassMetaData* 
SpindleBaseClass::getMetaData() const { 
  return &SpindleBaseClass_MetaData;
}

SpindleBaseClass::~SpindleBaseClass() {
  decrementInstanceCount( SpindleBaseClass::MetaData );
}

SpindleBaseClass::SpindleBaseClass() : idNumber(UniqueIDNumber++) {  
  incrementInstanceCount( SpindleBaseClass::MetaData );
}

SpindleBaseClass::SpindleBaseClass( const SpindleBaseClass& src ) : idNumber(UniqueIDNumber++) {
  // dummy use of src to keep compiler from complaining
  (void) src;
}

void
SpindleBaseClass::operator=( const SpindleBaseClass& src ) {
  // dummy use of src to keep compiler from complaining
  (void) src;
}

#define __FUNC__ "void SpindleBaseClass::loadObject( SpindleArchive& ar )"
void
SpindleBaseClass::loadObject( SpindleArchive& ar ) {
  FENTER;
  ASSERT( false, "Method not implemented.");
# ifndef _MSC_VER
  (void) ar;
# endif
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "void SpindleBaseClass::loadObject( SpindleArchive& ar )"
void
SpindleBaseClass::storeObject( SpindleArchive& ar ) const {
  FENTER;
  ASSERT( false, "Method not implemented.");
# ifndef _MSC_VER
  (void) ar;
# endif
  FEXIT;
}
#undef __FUNC__

// bool 
// SpindleBaseClass::isKindOf( const ClassMetaData* metaData ) const { 
//   const ClassMetaData* myType = this->getMetaData();
//   for(const ClassMetaData* temp = metaData; temp != 0; temp = temp->parentMetaData) {
//     if ( temp == myType ) {
// 	 return true;
//     } 
//   }
//   return false;
// }

SpindleBaseClass* 
SpindleBaseClass::dynamicCast( SpindleBaseClass * object ) {
  if ( object == 0 ) { return 0; }
  else if ( object->getMetaData() != SpindleBaseClass::MetaData ) { return 0; }
  else { return (SpindleBaseClass *) object; }
}

const SpindleBaseClass* 
SpindleBaseClass::dynamicCast( const SpindleBaseClass * object ) {
  if ( object == 0 ) { return 0; }
  else if ( object->getMetaData() != SpindleBaseClass::MetaData ) { return 0; }
  else { return (const SpindleBaseClass *) object; }
}

SpindleBaseClass* 
SpindleBaseClass::createObject() 
{ return new SpindleBaseClass; }

void 
SpindleBaseClass::dump( FILE * stream ) const {
  fprintf(stream,"SpindleBaseClass:\n");
}
