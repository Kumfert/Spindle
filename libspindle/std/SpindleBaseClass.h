//
// SpindleBaseClass.h  -- A cosmic base class for the Spindle framework
//
//  $Id: SpindleBaseClass.h,v 1.2 2000/02/18 01:31:59 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
// =========================================================================
//
//


#ifndef SPINDLE_BASE_CLASS_H_
#define SPINDLE_BASE_CLASS_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_CLASS_META_DATA_H_
#include "spindle/ClassMetaData.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class SpindleArchive;

/**
 *  @memo Cosmic Abstract Base Class 
 *  @type class
 *
 * All classes derived from this have an overhead of four virtual
 * functions and a static ClassMetaData.
 *
 *  @author Gary Kumfert
 *  @version #$Id: SpindleBaseClass.h,v 1.2 2000/02/18 01:31:59 kumfert Exp $#
 *  @see ClassMetaData
 * 
 */
class SpindleBaseClass {
private:
  /// default copy constructor disabled, no implementation
  SpindleBaseClass( const SpindleBaseClass& src );      // no implementation
  /// default assignment operator disabled, no implementation
  void operator=( const SpindleBaseClass& src );      // no implementation
  
  /** 
   *  A unique identification number is assigned to each instance that
   *  inherits from SpindleBaseClass
   */
  int idNumber;

protected:
  /** 
   *  @memo the default constructor is protected 
   *  to control instantiation.
   */
  SpindleBaseClass();

  /**
   * This member function should be called by every constructor 
   *      of every class derived from #SpindleBaseClass#.
   * @memo increase the instance count in the #ClassMetaData# instance
   *       associated with the specific descendant of #SpindleBaseClass#
   * @param metaData must be a valid pointer
   */
  void incrementInstanceCount( const ClassMetaData * metaData );
  
  /**
   * This member function should be called by every constructor 
   *      of every class derived from #SpindleBaseClass#.
   * @memo decrease the instance count in the #ClassMetaData# instance
   *       associated with the specific descendant of #SpindleBaseClass#
   * @param metaData must be a valid pointer
   */
  void decrementInstanceCount( const ClassMetaData * metaData );

public:
  /** @memo destructor */
  virtual ~SpindleBaseClass();

  /**
   *  @memo get the unique ID number of this instance
   *  @return positive integer that is unique among all
   *          instances that inherit from SpindleBaseClass
   */
  int queryIDNumber() const { return idNumber; }

  /// get access to the meta data of a class
  virtual const ClassMetaData* getMetaData() const;

  /// dump the instance to a stream (usually for debugging)
  virtual void dump( FILE * stream ) const;

  /// load an instance from an archive (children of SpindlePersistant only!!!)
  virtual void loadObject( SpindleArchive& ar );

  /// write this instance to an archive (children of SpindlePersistant only!!!)
  virtual void storeObject( SpindleArchive& ar ) const;

  /// static pointer to metadata for each instance
  static const ClassMetaData* MetaData;

  /** 
   * used when #dynamic_cast<>()# was not available
   * @return the pointer only if it is safely down-casted from 
   *          #SpindleBaseClass#, NULL otherwise
   */
  static SpindleBaseClass* dynamicCast( SpindleBaseClass* object );

  /// used when #dynamic_cast<>()# was not available
  static const SpindleBaseClass* dynamicCast( const SpindleBaseClass* object );
  
  /// create an empty instance via static function
  static SpindleBaseClass* createObject();
};

inline void
SpindleBaseClass::incrementInstanceCount( const ClassMetaData* metaData ) {
  ClassMetaData* thisData = (ClassMetaData*) metaData; // cast away const-ness
  if ( thisData != 0 ) {
#ifdef SPINDLE__NO_INSTANCE_COUNT
    ;
#else
    thisData->nCreated++;
#endif
  }
}

inline void
SpindleBaseClass::decrementInstanceCount( const ClassMetaData* metaData ) {
  ClassMetaData* thisData = (ClassMetaData*) metaData;
  if ( thisData != 0 ) {
#ifdef SPINDLE__NO_INSTANCE_COUNT
    ;
#else
    thisData->nDestroyed++;
#endif
  }

}

class registry_init {
public:
  registry_init( const ClassMetaData* metaData );
};

#define META_DATA( thisClass ) ( ( ClassMetaData*)( thisClass::MetaData ) )

#define SPINDLE_DECLARE_DYNAMIC( thisClass ) \
public: \
  static const ClassMetaData* MetaData; \
  virtual const ClassMetaData* getMetaData() const; \
  static thisClass* dynamicCast( SpindleBaseClass* object); \
  static const thisClass* dynamicCast( const SpindleBaseClass* object); \

#define SPINDLE_IMPLEMENT_DYNAMIC( thisClass, parentClass ) \
  static const ClassMetaData thisClass##_MetaData = { \
    #thisClass, \
    sizeof( class thisClass ), \
    META_DATA( parentClass ), \
    0, 0, 0 \
    }; \
  const ClassMetaData* thisClass::MetaData = &thisClass##_MetaData; \
  static const registry_init init_##thisClass( &thisClass##_MetaData ); \
  const ClassMetaData* thisClass::getMetaData() const \
    { return  &thisClass##_MetaData ; } \
  thisClass* thisClass::dynamicCast( SpindleBaseClass* object) \
    { if ( object == 0 ) { return 0; } \
      else if (object->getMetaData() != thisClass::MetaData ) { return 0; }\
      else { return (thisClass *) object; } } \
  const thisClass* thisClass::dynamicCast( const SpindleBaseClass* object) \
    { if ( object == 0 ) { return 0; } \
      else if (object->getMetaData() != thisClass::MetaData ) { return 0; } \
      else { return (thisClass *) object; } } 

#define _SPINDLE_IMPLEMENT_DYNAMIC( thisClass, parentClass, pfNew ) \
  static const ClassMetaData thisClass##_MetaData = { \
    #thisClass, \
    sizeof( class thisClass ), \
    META_DATA( parentClass ), \
    pfNew, 0, 0 \
    }; \
  const ClassMetaData* thisClass::MetaData = & thisClass##_MetaData;\
  static const registry_init init_##thisClass( &thisClass##_MetaData ); \
  const ClassMetaData* thisClass::getMetaData() const \
    { return & thisClass##_MetaData ; } \
  thisClass* thisClass::dynamicCast( SpindleBaseClass* object) \
    { if ( object == 0 ) { return 0; } \
      else if (object->getMetaData() != thisClass::MetaData ) { return 0; } \
      else { return (thisClass *) object; } } \
  const thisClass* thisClass::dynamicCast( const SpindleBaseClass* object) \
    { if ( object == 0 ) { return 0; } \
      else if (object->getMetaData() != thisClass::MetaData ) { return 0; }\
      else { return (thisClass *) object; } } 

SPINDLE_END_NAMESPACE

#endif
