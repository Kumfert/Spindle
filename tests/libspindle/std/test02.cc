//
// src/std/drivers/tests/test02.cc
//
// $Id: test02.cc,v 1.2 2000/02/18 01:32:10 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  test Spindle Base class...
//

#include "spindle/SpindleBaseClass.h"
#include <iostream.h>

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

class Test1 : public SpindleBaseClass {
  SPINDLE_DECLARE_DYNAMIC( Test1 )
protected:
  int someInt;
  Test1() { someInt=0; }
public:
  Test1( int newInt ) : someInt(newInt) {}
  int getData() const { return someInt; }
  void setData(int newInt) { someInt = newInt; }
  virtual void dump(FILE* fp) const;
};

SPINDLE_IMPLEMENT_DYNAMIC( Test1, SpindleBaseClass )

void
Test1::dump( FILE* fp ) const {
  //  ::dump(fp);
  fprintf(fp,"Test1::someInt = %d\n", someInt);
}

class Test2 : public Test1 {
  SPINDLE_DECLARE_DYNAMIC( Test2 )
protected:
  int anotherInt;
  Test2() { anotherInt=2; }
public:
  Test2( int newInt ) : anotherInt(newInt) {}
  int getOtherData() const { return anotherInt; }
  void setOtherData(int newInt ) { anotherInt = newInt; }
  virtual void dump( FILE* fp ) const;
};

SPINDLE_IMPLEMENT_DYNAMIC( Test2, Test1 )

void
Test2::dump( FILE* fp ) const {
  // ::dump(fp);
  fprintf(fp,"Test2::someInt = %d\nTest2::anotherInt = %d\n", someInt, anotherInt);
}



class Exam1 : public SpindleBaseClass {
  SPINDLE_DECLARE_DYNAMIC( Exam1 )
private:
  float someFloat;
  Exam1() { someFloat=0.0; }
public:
  Exam1( float newFloat ) : someFloat( newFloat ) {}
  float getData() const { return someFloat; }
  void setData(float newFloat) { someFloat = newFloat; }
  virtual void dump(FILE* fp) const;
};

SPINDLE_IMPLEMENT_DYNAMIC( Exam1, SpindleBaseClass )

void
Exam1::dump( FILE* fp ) const {
  //  ::dump(fp);
  fprintf(fp,"Exam1::someFloat = %f\n", someFloat);
}


int main() {
  Test1 test1(5);
  const SpindleBaseClass* t1 = &test1;

  cout << "Is *t1 a kind of SpindleBaseClass ?   " << flush;
  const SpindleBaseClass* pBC;
  if ( ( pBC = SpindleBaseClass::dynamicCast( t1 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  cout << "Is *t1 a kind of Test1 ?            " << flush;
  const Test1* pTest1;
  if ( ( pTest1 = Test1::dynamicCast( t1 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  cout << "Is *t1 a kind of Test2 ?            " << flush;
  const Test2* pTest2;
  if ( ( pTest2 = Test2::dynamicCast( t1 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }
  
  cout << "Is *t1 a kind of Exam1 ?            " << flush;
  const Exam1 *pExam1;
  if ( ( pExam1 = Exam1::dynamicCast( t1 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  t1->dump( stdout );

  Test2 test2(5);
  test2.setData(1);
  const SpindleBaseClass* t2 = &test2;

  cout << "Is *t2 a kind of SpindleBaseClass ?   " << flush;
  if ( (pBC = SpindleBaseClass::dynamicCast( t2 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  cout << "Is *t2 a kind of Test1 ?            " << flush;
  if ( (pTest1 = Test1::dynamicCast( t2 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  cout << "Is *t2 a kind of Test2 ?            " << flush;
  if ( (pTest2 = Test2::dynamicCast( t2 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  cout << "Is *t2 a kind of Exam1 ?            " << flush;
  if ( ( pExam1 = Exam1::dynamicCast( t2 ) ) != 0 ) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }

  t2->dump( stdout );


}
