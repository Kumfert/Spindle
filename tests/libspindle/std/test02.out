Is *t1 a kind of SpindleBaseClass ?   NO
Is *t1 a kind of Test1 ?            YES
Is *t1 a kind of Test2 ?            NO
Is *t1 a kind of Exam1 ?            NO
Test1::someInt = 5
Is *t2 a kind of SpindleBaseClass ?   NO
Is *t2 a kind of Test1 ?            NO
Is *t2 a kind of Test2 ?            YES
Is *t2 a kind of Exam1 ?            NO
Test2::someInt = 1
Test2::anotherInt = 5
